docs/proposals/role_decomposition.md

```diff
@@ -0,0 +1,322 @@
+# Scaffolding for decomposing large roles
+
+## Why?
+
+Currently we have roles that are very large and encompass a lot of different
+components. This makes for a lot of logic required within the role, can
+create complex conditionals, and increases the learning curve for the role.
+
+## How?
+
+Creating a guide on how to approach breaking up a large role into smaller,
+component based, roles. Also describe how to develop new roles, to avoid creating
+large roles.
+
+## Proposal
+
+Create a new guide or append to the current contributing guide a process for
+identifying large roles that can be split up, and how to compose smaller roles
+going forward.
+
+### Large roles
+
+A role should be considered for decomposition if it:
+
+1) Configures/installs more than one product.
+1) Can configure multiple variations of the same product that can live
+side by side.
+1) Has different entry points for upgrading and installing a product
```

Comment:
I agree, this may be a "pie in the sky" goal and we may not be able to achieve it with all of our roles. Some of our hosted roles like Logging and Metrics will likely be able to follow this, but for `openshift_master` might work better to be a playbook to orchestrate work flow (e.g. mark a node `unschedulable`, pull it out of a HA cluster, etc) and then usesthe smaller component roles to install and configure the parts that make it up